text/0000-wallet-replay-protection.md

```diff
@@ -0,0 +1,322 @@
+2. A `PayJoin` tx which has inputs from both parties - both sender and receiver contribute at least one input each
+
+Let's call outputs that hang from a chain that leads to your anchor a `Protected` output. A transaction is protected from replays if a party contributes a Protected output as an input in the transaction. It's not enough that some other party contributes such an input. We are only really protected from replay attacks if we contribute the `Protected` output. Otherwise, our security depends on the honesty and following of the wallet rules of other participants which should not be relied on.
+
+Let's say that we want to be able to do both `Regular` and `PayJoin` transactions. We do both of these so we have some protected outputs and some that are not protected. The first problem we have is that after a wallet restore, we can't really tell whether an output is protected through a PayJoins+Anchor or not. It would be nice if we could label our outputs as either `Protected` or `Unprotected` after a wallet restore.
+
+### Separation of replay protected outputs
+
+Let's define a clear separation of outputs that are protected from those that are not. We define a protected output generator `GenP` that allows us to either `create` a protected output or `check` whether an output is `Protected`.
+
+An output is labeled as `Protected` if it was created in a transaction where we either contributed a protected input, or contributed an anchor output.
+
+#### Possible protected output generators implementations
+
+As we mentioned above, `GenP` for `Protected` outputs would need to have `create` and `check` functions defined.
+
+There are a few choices how to label outputs as `Protected` while still being able to identify them across difference devices:
+1. Call to `create` uses a new derivation path `P` that is used only for creating `Protected` outputs. Similarly `check` uses the same `P` to check whether an output is protected. Perhaps we could have `N` labels possible which would be labeled by the `r % N` result
+2. Call to `create` creates a specific `r` value for `Protected` outputs e.g. they should start with `N` zeros or be divisible by some number `M`
+3. Call to `create` uses additional output information in the bytes that are available in the Bulletproofs to convey the idea whether the output is protected. These bytes are right now 'zero' bytes meaning their binary representation is all zeros. We label an output as `Protected` by setting the `label` bit that is unused now to `1`. `Unprotected` outputs will have the label bit set to `0` which includes all the old outputs as well. To make sure the output label would be seen only to the owner we save a xor `bit = label ^ r`. This way, nobody would know which label the output has except for the owner of the utxo because they know the `r` value. The owner could xor again to get the label `label = bit ^ r`. If we went such path, we would need to think of the possible drawbacks.

```

Comment:
I think all the data recoverable from the Bulletproof is already private, so there's no need to do any xoring.
Also, label^r makes no sense since r is 256 bits.

Reply:
thanks, I didn't know that. My bad, I was thinking xoring the least significant bit of r, but if it's already private then it's not even a problem.